Zend/zend_execute.c

```diff
@@ -2642,6 +2708,23 @@ static zend_never_inline zend_bool zend_handle_fetch_obj_flags(
 				ZEND_REF_ADD_TYPE_SOURCE(Z_REF_P(ptr), prop_info);
 			}
 			break;
+		case ZEND_FETCH_DIM_UNSET:
+		case ZEND_FETCH_DIM_RW:
+			if (!prop_info) {
+				prop_info = zend_object_fetch_property_type_info(obj, ptr);
+				if (!prop_info) {
+					break;
+				}
+			}
+
+			if (prop_info->flags & ZEND_ACC_FINAL && Z_PROP_FLAG_P(ptr) != IS_PROP_UNINIT) {
+				zend_throw_error(NULL, "Cannot modify final property %s::$%s after initialization",
+					ZSTR_VAL(prop_info->ce->name), zend_get_unmangled_property_name(prop_info->name)
+				);
+				if (result) ZVAL_ERROR(result);
+				return 0;
+			}
+			break;
```

Comment:
Does this code actually do anything? I don't see any corresponding compiler code that sets these flags, so I would expect it to never run.

Reply:
No, not much. I put it as an earlier experimentation (without checking the compiler), but forgot to remove it.

Reply:
Hmm, or is it possible that I'll need this code? Currently, I'm trying to implement the restriction for the `ZEND_FETCH_OBJ_UNSET` opcode.  I see that `zend_fetch_property_address()` is called from the opcode handler with `flags=0`. However, should I have this piece of code and set the flag then maybe `unset($foo->a[0])` could be checked? Or is there a better way?

Reply:
Yes, I think you're going to need at least the UNSET portion of this, in order to distinguish between `unset($foo->a[0])` and `unset($foo->a->b)` (otherwise you could just check for BP_VAR_UNSET). You'll have to add the corresponding flag in the compiler (like it is done for the two flags that are checked above). One value was freed up in PHP 8, so it should fit in.